tests/testthat/test-dfm_group.R

```diff
@@ -56,3 +56,94 @@ test_that("dfm.character groups works (#794)", {
         dfm(corp, groups = "grp")
     )
 })
+
+test_that("test dfm_group with factor levels, fill = TRUE and FALSE, #854", {
+    
+    corp <- corpus(c('a b c c', 'b c d', 'a'),
+                   docvars = data.frame(grp = factor(c("A", "A", "B"), levels = LETTERS[1:4])))
+    testdfm <- dfm(corp)
+    expect_equal(
+        as.matrix(dfm_group(testdfm, groups = "grp", fill = FALSE)),
+        matrix(c(1,2,3,1, 1,0,0,0), byrow = TRUE, nrow = 2, 
+               dimnames = list(docs = c("A", "B"), features = letters[1:4]))
+    )
+    expect_equal(
+        as.matrix(dfm_group(testdfm, groups = "grp", fill = TRUE)),
+        matrix(c(1,2,3,1, 1,0,0,0, 0,0,0,0), byrow = TRUE, nrow = 3, 
+               dimnames = list(docs = c("A", "B", "C"), features = letters[1:4]))
+    )
+    
+    testdfm <- dfm(c('a b c c', 'b c d', 'a'))
+    external_factor <- factor(c('doc1', 'doc1', 'doc2'), 
+                              levels = c('doc0', 'doc1', 'doc2', 'doc3'))
+    expect_equal(
+        as.matrix(dfm_group(testdfm, groups = external_factor, fill = FALSE)),
+        matrix(c(1,2,3,1, 1,0,0,0), byrow = TRUE, nrow = 2, 
+               dimnames = list(docs = c('doc1', 'doc2'), features = letters[1:4]))
+    )
+    expect_equal(
+        as.matrix(dfm_group(testdfm, groups = external_factor, fill = TRUE)),
+        matrix(c(0,0,0,0, 1,2,3,1, 1,0,0,0, 0,0,0,0), byrow = TRUE, nrow = 4, 
+               dimnames = list(docs = paste0("doc", 0:3), features = letters[1:4]))
+    )
+    # new documents in factor order
+    expect_equal(
+        as.matrix(dfm_group(testdfm, groups = factor(c(1, 1, 2), levels = 4:1), fill = TRUE)),
+        matrix(c(rep(0, 8), 1,0,0,0, 1,2,3,1), byrow = TRUE, nrow = 4,
+               dimnames = list(docs = 4:1, features = letters[1:4]))
+    )
+    # should this also be ordered? (here the expectation is that it is)
+    expect_equal(
+        as.matrix(dfm_group(testdfm, groups = factor(c(3, 3, 1), levels = 4:1), fill = FALSE)),
+        matrix(c(1,2,3,1, 1,0,0,0), byrow = TRUE, nrow = 4,
+               dimnames = list(docs = c(3, 1), features = letters[1:4]))
+    )
+})
+
+test_that("test dfm_group with non-factor grouping variable, with fill", {
+    grp <- c("D", "D", "A", "C")
+    corp <- corpus(c('a b c c', 'b c d', 'a', "b d d"),
+                   docvars = data.frame(grp = grp, stringsAsFactors = FALSE))
+    testdfm <- dfm(corp)
+    expect_equal(
+        as.matrix(dfm_group(testdfm, groups = "grp", fill = FALSE)),
+        matrix(c(1,0,0,0, 0,1,0,2, 1,2,3,1), byrow = TRUE, nrow = 3, 
+               dimnames = list(docs = c("A", "B", "D"), features = letters[1:4]))
```

Comment:
Oops, this should be `docs = c("A", "B", "D")` ...  But it's an open question as to whether this should be ordered or not. My view is that they should, since we can't assume there is any particular order in the groups assigned to the documents, simply in the way that they are assigned.